platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -265,535 +292,130 @@ public Hash hash(
-                lq.stream().forEach(job -> q.appendHashJob().dirtyLeaf(job.getPath(), job.getLeaf()));
-                lq = null;
+                curPath = parentPath;
+                curTask = parentTask;
             }
-
-            // Reset the pending queue and now initialOffset will be zero for all subsequent ranks.
-            pq.reset();
+            firstLeaf = false;
         }
-    }
-
-    /**
-     * Read all leaves from the given iterator that are in the given rank and segment and add them to the given buffer.
-     *
-     * @param itr
-     * 		The iterator to read from. Cannot be null. May be at the end.
-     * @param queue
-     * 		The queue to write leaves into. The queue <strong>will be</strong> reset.
-     * @param eofPath
-     * 		The path that is one place larger than the last valid path to read to.
-     * 		This could be the end of a rank, or the end of a specific segment.
-     */
-    private void readLeavesInSegment(
-            final PeekIterator<VirtualLeafRecord<K, V>> itr, final HashingQueue<K, V> queue, final long eofPath) {
-        // Iterate either until we run out of dirty leaves, or we encounter a leaf that is in the wrong rank or segment
-        while (itr.hasNext()) {
-            // Peek at the next leaf, but don't pull it off until we determine that it belongs in this buffer
-            final VirtualLeafRecord<?, ?> nextLeaf = itr.peek();
-            final long path = nextLeaf.getPath();
-            if (path < eofPath) {
-                // We know that this leaf belongs in the buffer, so we can go ahead and read it off now.
-                queue.appendHashJob().dirtyLeaf(path, itr.next());
-            } else {
-                // We have encountered a leaf that is in the wrong segment.
-                break;
+        map.forEach((path, task) -> task.setData(null));
+        map.clear();
+
+        try {
+            rootTask.join();
+        } catch (final Exception e) {
+            if (shutdown.get()) {
+                return null;
             }
+            logger.error(EXCEPTION.getMarker(), "Failed to wait for all hashing tasks", e);
+            throw e;
         }
-    }
 
-    private void accumulate(
-            final PeekIterator<VirtualLeafRecord<K, V>> itr,
-            final HashingQueue<K, V> queue,
-            long segmentStart,
-            final long segmentSize,
-            final long eofPath) {
-        // While we have not yet read our "preferred" number of items, read off a segment into the queue.
-        while (queue.size() < segmentSize) {
-            readLeavesInSegment(itr, queue, Math.min(segmentStart + segmentSize, eofPath));
-            // If we have read as much as is available to read, then bail.
-            if (!itr.hasNext() || itr.peek().getPath() >= eofPath) {
-                return;
-            } else {
-                segmentStart += segmentSize;
-            }
-        }
+        listener.onHashingCompleted();
```

Comment:
Minor: should this be in a sort of a `finally{}` block of some kind to ensure the listener's state isn't corrupt in case the above code throws an exception or otherwise fails the task?

Reply:
This is no different from the current implementation. My observation is if hashing fails, this is an unrecoverable error, and there is no need to call `onHashingCompleted()`

Reply:
Oh, I see. Sounds good then.